drivers/ethernet/eth_xlnx_gem.c

```diff
@@ -0,0 +1,1733 @@
+	dev_data->txbd_ring.next_to_use =
+		(first_bd_idx + bds_reqd) % dev_conf->txbd_count;
+	dev_data->txbd_ring.free_bds -= bds_reqd;
+
+	if (dev_conf->defer_txd_to_queue) {
+		k_sem_give(&(dev_data->txbd_ring.ring_sem));
+	} else {
+		sys_write32(ETH_XLNX_GEM_IXR_TX_COMPLETE_BIT,
+			dev_conf->base_addr + ETH_XLNX_GEM_IER_OFFSET);
+	}
+
+	/*
+	 * Scatter the contents of the network packet's buffer to
+	 * one or more DMA buffers.
+	 */
+	net_pkt_cursor_init(pkt);
+	do {
+		/* Calculate the base pointer of the target TX buffer */
+		tx_buffer_offs = (void *)(dev_data->first_tx_buffer +
+			(dev_conf->tx_buffer_size * curr_bd_idx));
+
+		/* Copy packet data to DMA buffer */
+		net_pkt_read(
+			pkt, (void *)tx_buffer_offs,
+			(tx_data_remaining < dev_conf->tx_buffer_size) ?
+			tx_data_remaining : dev_conf->tx_buffer_size);
+
+		/* Update current BD's control word */
+		reg_val = sys_read32(
+			(uint32_t)(&dev_data->txbd_ring.first_bd[curr_bd_idx].ctrl)) &
+			(ETH_XLNX_GEM_TXBD_WRAP_BIT | ETH_XLNX_GEM_TXBD_USED_BIT);
+		reg_val |= (tx_data_remaining < dev_conf->tx_buffer_size) ?
+			tx_data_remaining : dev_conf->tx_buffer_size;
+		sys_write32(reg_val,
+			(uint32_t)(&dev_data->txbd_ring.first_bd[curr_bd_idx].ctrl));
+
+		if (tx_data_remaining > dev_conf->tx_buffer_size) {
+			/* Switch to next BD */
+			curr_bd_idx = (curr_bd_idx + 1) % dev_conf->txbd_count;
+		}
+
+		tx_data_remaining -=
+			(tx_data_remaining < dev_conf->tx_buffer_size) ?
+			tx_data_remaining : dev_conf->tx_buffer_size;
+	} while (tx_data_remaining > 0);
+
+	/* Set the 'last' bit in the current BD's control word */
+	reg_val |= ETH_XLNX_GEM_TXBD_LAST_BIT;
+
+	/*
+	 * Clear the 'used' bits of all BDs involved in the current
+	 * transmission. In accordance with chapter 16.3.8 of the
+	 * Zynq-7000 TRM, the 'used' bits shall be cleared in reverse
+	 * order, so that the 'used' bit of the first BD is cleared
+	 * last just before the transmission is started.
+	 */
+	reg_val &= ~ETH_XLNX_GEM_TXBD_USED_BIT;
+	sys_write32(reg_val,
+		(uint32_t)(&dev_data->txbd_ring.first_bd[curr_bd_idx].ctrl));
+
+	while (curr_bd_idx != first_bd_idx) {
+		curr_bd_idx = (curr_bd_idx != 0) ?
+			(curr_bd_idx - 1) : (dev_conf->txbd_count - 1);
+		reg_val = sys_read32(
+			(uint32_t)(&dev_data->txbd_ring.first_bd[curr_bd_idx].ctrl));
+		reg_val &= ~ETH_XLNX_GEM_TXBD_USED_BIT;
+		sys_write32(reg_val,
+			(uint32_t)(&dev_data->txbd_ring.first_bd[curr_bd_idx].ctrl));
+	}
+
+	/* Set the start TX bit in the gem.net_ctrl register */
+	reg_val = sys_read32(
+		dev_conf->base_addr + ETH_XLNX_GEM_NWCTRL_OFFSET);
+	reg_val |= ETH_XLNX_GEM_NWCTRL_STARTTX_BIT;
+	sys_write32(reg_val,
+		dev_conf->base_addr + ETH_XLNX_GEM_NWCTRL_OFFSET);
+
+#ifdef CONFIG_NET_STATISTICS_ETHERNET
+	dev_data->stats.bytes.sent += tx_data_length;
+	dev_data->stats.pkts.tx++;
+#endif
+
+	/* Block until TX has completed */
+	sem_status = k_sem_take(&dev_data->tx_done_sem, K_MSEC(5000));
```

Comment:
The timeout is quite long, is it really plausible that we could wait this long. Could we make this timeout shorter?